src/execution_plan/ops/op_delete.c

```diff
@@ -201,38 +199,69 @@ static Record DeleteConsume(OpBase *opBase) {
 			break;
 		}
 	}
-
-	return r;
 }
 
-static OpBase *DeleteClone(const ExecutionPlan *plan, const OpBase *opBase) {
-	ASSERT(opBase->type == OPType_DELETE);
-	OpDelete *op = (OpDelete *)opBase;
-	AR_ExpNode **exps;
-	array_clone_with_cb(exps, op->exps, AR_EXP_Clone);
-	return NewDeleteOp(plan, exps);
+// Return mode, emit a populated Record.
+static Record _handoff(OpDelete *op) {
+	return array_pop(op->records);
 }
 
-static OpResult DeleteReset(OpBase *opBase) {
+static Record DeleteConsume(OpBase *opBase) {
 	OpDelete *op = (OpDelete *)opBase;
+	Record r;
 
-	_DeleteEntities(op);
+	// Return mode, all data was consumed.
+	if(op->records) return _handoff(op);
 
-	if(op->deleted_nodes) {
-		array_clear(op->deleted_nodes);
-	}
+	// Consume mode.
+	op->records = array_new(Record, 32);
+	// initialize the records array with NULL, which will terminate execution
+	// upon depletion
+	array_append(op->records, NULL);
 
-	if(op->deleted_edges) {
-		array_clear(op->deleted_edges);
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	ASSERT(op->op.childCount > 0);
+	// Pull data until child is depleted.
+	OpBase *child = op->op.children[0];
+	while((r = OpBase_Consume(child))) {
+		/* Persist scalars from previous ops before storing the record,
+		* as those ops will be freed before the records are handed off. */
+		Record_PersistScalars(r);
+
+		// save record for later use
+		array_append(op->records, r);
+
+		// delete entities
+		_CollectDeletedEntities(r, opBase);
 	}
 
-	return OP_OK;
+	/* Done reading, we're not going to call consume any longer
+	 * there might be operations e.g. index scan that need to free
+	 * index R/W lock, as such free all execution plan operation up the chain. */
+	if(child) OpBase_PropagateReset(child);
```

Comment:
Looking a head, I think a UNION queries can introduce issues, consider the following plan:
```
1) "Results"
2) "    Join"
3) "        Project"
4) "            Project"
5) "                Delete"
6) "                    Node By Index Scan | (n:N)"
7) "        Project"
8) "            Node By Index Scan | (n:N)"
```

I'm not saying we should address this issue in this PR.

Reply:
???

Reply:
The delete operation will send a `reset` request to free all Search indices but this reset request will not reach the IndexScan operation on the other side of the Join op.

Reply:
Join reset the streams that finished this was fixed lately